### PR TITLE
feat: Add ability to configure environments, collaborators, and variables

### PR DESCRIPTION
**Pull request by actuarysailor**
## Description
Updated the action to enable the configuration of:
- Deployment Environments #218, including:
  - Deployment Protection Policies
  - Variables
  - Secrets
- Collaborators, both teams and users as well as their permissions #30
- Repository variables #219

## Motivation and Context
I manage several hundred repositories at work, and we use templates to ensure that everyone has the correct CI/CD pipelines for various project types varying by platform, programming language, and teams responsible for them.  While this action does not fully automate everything such that all repositories require no manual configurations, it does streamline the configuration process for new repositories, such that:
- Baseline permissions for contributors get automatically set
- Baseline configurations for common settings get automatically set

While new repositories still need their settings.yml file updated for any customizations specific to a new repository, this allows even non-admins to propose those updates via a pull request.  Secret values may still need a manual configuration because it is not guaranteed that:
- the worker running this action would have the appropriate values stored in an environment variable
- hard-coding secret values into the settings.yml file directly defeats the purpose of having an encrypted repo secret

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects

### CODE REVIEW

**Comment by actuarysailor** [comment]
@andrewthetechie  - I would really like to get this one completed.  Greatly enhances the features of your action to fully manage nearly everything in a repository configuration

**Comment by andrewthetechie** [comment]
@actuarysailor I'll take a look this week and try to get CI passing

**Comment by actuarysailor** [comment]
@andrewthetechie - yes, I tried to fix as many of the CI issues as I could, the few remaining I’m not understanding what your checking for.  Also seems like the issue may be in sections of code I did not update.

File: pyproject.toml [hunk 1]
```diff
@@ -13,7 +13,8 @@ repo-manager = "repo_manager.main:main"
 
 [tool.poetry.dependencies]
 python = "^3.11"
-pydantic = "^1.10.9"
+pydantic = "^2.7.1"
```
Comment (andrewthetechie) [inline pyproject.toml:16]: pydantic 2 is a major breaking change and all the schemas would have to be updated to support it.

If you want to do that, I support it, but it should be in its own PR and all schemas updated to use the new v2 syntax

File: repo_manager/github/environments.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+    return status == 204
+
+
+def diff_option(key: str, expected: Any, repo_value: Any) -> str | None:
+    if expected != repo_value:
+        return f"{key} -- Expected: {expected} Found: {repo_value}"
+    return None
+
+
+def create_environment(repo: Repository, env: environment) -> bool:
```
Comment (andrewthetechie) [inline repo_manager/github/environments.py:141]: pygithub has support to manage Environments. Can we use that instead of maintaining code in the action?

File: repo_manager/github/variables.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+def _get_repo_variable_names(repo: Repository, path: str = "actions") -> set[str]:
+    return {variable["name"] for variable in _get_repo_variables(repo, path)}
+
+
+def _get_repo_variable_dict(repo: Repository, path: str = "actions") -> dict[str, Any]:
+    return {variable["name"]: variable for variable in _get_repo_variables(repo, path)}
+
+
+def create_variable(
```
Comment (andrewthetechie) [inline repo_manager/github/variables.py:42]: pygithub has support for managing variables, can that be used instead of re-implementing the API calls in the action?

File: repo_manager/schemas/environment.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    @model_validator(mode="after")
+    def initialize_branch_name_patterns(self) -> Self:
+        if self.protected_branches == self.custom_branch_policies:
+            raise ValueError(
+                "You must specify either protected branches or custom branch policies, not both."
+            )
+        return self
+
+
+class environment(BaseModel):
```
Comment (andrewthetechie) [inline repo_manager/schemas/environment.py:69]: class should have a capitalized first letter in its name

File: repo_manager/utils/__init__.py [hunk 5]
```diff
@@ -90,6 +90,13 @@ def validate_inputs(parsed_inputs: dict[str, Any]) -> dict[str, Any]:
 
     actions_toolkit.debug(f"api_url: {api_url}")
 
+    try:
```
Comment (andrewthetechie) [inline repo_manager/utils/__init__.py:93]: Why does this need to be duplicated? it seems to duplicate the lines directly below it

**Review by andrewthetechie — CHANGES_REQUESTED** [review]
@actuarysailor I think these changes are great features, and I'd be happy to merge them in. Thank you so much for taking the time to implement this. It's very cool that this action has been helpful to you. I built it to solve a problem for myself (not wanting to pay for a GitHub org for myself :P), but I'm glad it's been useful outside of that.

Before merging these changes, I left some comments on some code changes. I also ask that these features and changes be split into separate PRs instead of one large PR. Reviewing smaller PRs and reverting will be much easier if there are problems.


This should end up being 4 PRs:

* pydantic 2.0 bump and schema changes
* PR to implement Variables
* PR to implement collaborators
* PR to implement Deployment environments

Wherever possible, try to use the upstream pygithub methods for managing GitHub resources rather than re-implementing API calls. If you have to implement an API call, please document the missing feature from pygithub so that the code can be reviewed in the future to see if it can be moved to using pygithub.

Let me know if you have any questions on any of my comments! I'll check back several times during the week to answer questions or review smaller PRs as you have time to make them.

File: pyproject.toml (hunk 1 above)
Comment (actuarysailor) [reply pyproject.toml:16]: @andrewthetechie  see #225

**Comment by actuarysailor** [comment]
> @actuarysailor I think these changes are great features, and I'd be happy to merge them in. Thank you so much for taking the time to implement this. It's very cool that this action has been helpful to you. I built it to solve a problem for myself (not wanting to pay for a GitHub org for myself :P), but I'm glad it's been useful outside of that.
> 
> Before merging these changes, I left some comments on some code changes. I also ask that these features and changes be split into separate PRs instead of one large PR. Reviewing smaller PRs and reverting will be much easier if there are problems.
> 
> This should end up being 4 PRs:
> 
> * pydantic 2.0 bump and schema changes
> * PR to implement Variables
> * PR to implement collaborators
> * PR to implement Deployment environments
> 
> Wherever possible, try to use the upstream pygithub methods for managing GitHub resources rather than re-implementing API calls. If you have to implement an API call, please document the missing feature from pygithub so that the code can be reviewed in the future to see if it can be moved to using pygithub.
> 
> Let me know if you have any questions on any of my comments! I'll check back several times during the week to answer questions or review smaller PRs as you have time to make them.

@andrewthetechie - I will add the other PRs / clean them up per your suggestions (if possible); I know some of the GitHub methods have invalid URLs, but I also see you changed version on pygithub, so maybe it is fixed in the new version.

**Comment by actuarysailor** [comment]
@andrewthetechie - I changed to a draft since I am making the other PRs as per your request.

**Comment by andrewthetechie** [comment]
Closing these per discussion with @actuarysailor in another PR.
